experiment/simulation/js/gate.js

```diff
@@ -0,0 +1,317 @@
+        this.output = val;
+    }
+    setConnected(val) {
+        this.isConnected = val;
+    }
+}
+
+
+
+function addGate(event) {
+    const type = event.target.innerHTML;
+    const gate = new Gate(type);
+    const component = gate.generateComponent();
+    const parent = document.getElementById("working-area");
+    parent.insertAdjacentHTML('beforeend', component);
+    gate.registerComponent("working-area");
+}
+
+window.addGate = addGate;
+
+export function getResult(gate) {
+    if (gate.output != null) {
+        return;
+    }
+    for (let i = 0; i < gate.inputs.length; i++) {
+        if (gate.inputs[i].output == null) {
+            getResult(gate.inputs[i]);
+        }
+    }
+    gate.generateOutput();
+    return;
+}
+
+function setInput(event) {
+    let parentElement = event.target.parentElement;
+    let element = event.target;
+    let type = parentElement.className.split(" ")[0];
+    let gate = gates[parentElement.id];
+    if (type === "high") {
+        // change class HIGH to LOW
+        parentElement.classList.replace("high", "low");
+        element.innerHTML = "0";
+        gate.setOutput(false);
+    }
+    else if (type === "low") {
+        parentElement.classList.replace("low", "high");
+        element.innerHTML = "1";
+        gate.setOutput(true);
+    }
+}
+
+window.setInput = setInput;
+
+export function clearResult() {
+    const result = document.getElementById("result");
+    result.innerHTML = "";
+}
+
+export function printErrors(message,objectId) {
+    const result = document.getElementById('result');
+    result.innerHTML += message;
+    result.className = "failure-message";
+    if(objectId !== null)
+    {
+        objectId.classList.add("highlight")
+        setTimeout(function () {objectId.classList.remove("highlight")}, 5000);
+    }
+}
+
+// Check if the connections are correct
+export function checkConnections() {
+    for (let gateId in gates) {
+        const gate = gates[gateId];
+        const id = document.getElementById(gate.id);
+        if (gate.inputPoints.length != gate.inputs.length) {
+            printErrors("Highlighted component not connected properly\n",id);
+            return false;
+        } else if (gate.isConnected === false && gate.isOutput === false) {
+            printErrors("Highlighted component not connected properly\n",id);
+            return false;
+        }
+    }
+    return true;
+}
+
+// Simulate the circuit
+export function simulate() {
+    clearResult();
+    if (!checkConnections()) {
+        return;
+    }
+
+    // reset output in gate
+    for (let gateId in gates) {
+        if (!gates[gateId].isInput) {
+            gates[gateId].output = null;
+        }
+    }
+
+    for (let gateId in gates) {
+        const gate = gates[gateId];
+        if (gate.isOutput) {
+            getResult(gate);
+            let element = document.getElementById(gate.id);
+            if (gate.output) {
+                element.className = "high";

```

Comment:
Will it be better to add to the existing classnames instead of just assigning a classname, which will effectively remove all the previously existing classnames and may cause subtle bugs

Reply:
there are 3 classes, output, high and low. Output is the default class, high and low are the classes assigned according to its state. So directly assigning didn't make much difference as it had just one class assigned to itself.